core/src/typecheck/subtyping.rs

```diff
@@ -0,0 +1,238 @@
+//! Types subsumption.
+/// Change from inference mode to checking mode, and apply a potential subsumption rule.
+///
+/// Currently, there is record/dictionary subtyping, if we are not in this case we fallback to perform
+/// polymorphic type instantiation with unification variable on the left (on the inferred type),
+/// and then simply performs unification (put differently, the subtyping relation when it is not
+/// a record/dictionary subtyping is the equality
+/// relation).
+///
+/// The type instantiation corresponds to the zero-ary case of application in the current
+/// specification (which is based on [A Quick Look at Impredicativity][quick-look], although we
+/// currently don't support impredicative polymorphism).
+///
+/// In the future, this function might implement a other non-trivial subsumption rule.
+///
+/// [quick-look]: https://www.microsoft.com/en-us/research/uploads/prod/2020/01/quick-look-icfp20-fixed.pdf
+use super::*;
+
+pub(super) trait Subsume {
```

Comment:
In Rust, it's customary to give the same name to the trait and to the provided function, when a trait provides one single well scoped method.

```suggestion
pub(super) trait IsSubsumedBy {
```